pkg/tanka/export.go

```diff
@@ -0,0 +1,209 @@
+	Targets   []string  // optional: only export specified Kubernetes manifests
+	ParseOpts ParseOpts // optional: options for parsing Environments
+}
+
+func DefaultExportEnvOpts() ExportEnvOpts {
+	format := "{{env.spec.namespace}}/{{env.metadata.name}}/{{.apiVersion}}.{{.kind}}-{{.metadata.name}}"
+	extension := "yaml"
+	merge := false
+	return ExportEnvOpts{
+		Format:    &format,
+		Extension: &extension,
+		Merge:     &merge,
+	}
+}
+
+func ExportEnvironments(paths []string, to string, opts *ExportEnvOpts) error {
+	// Keep track of which file maps to which environment
+	fileToEnv := map[string]string{}
+
+	// dir must be empty
+	empty, err := dirEmpty(to)
+	if err != nil {
+		return fmt.Errorf("Checking target dir: %s", err)
+	}
+	if !empty && !*opts.Merge {
+		return fmt.Errorf("Output dir `%s` not empty. Pass --merge to ignore this", to)
+	}
+
+	// get all environments for paths
+	envs, err := ParseEnvs(paths, opts.ParseOpts)
+	if err != nil {
+		return err
+	}
+
+	for _, env := range envs {
+		// select targets to export
+		filter, err := StringsToRegexps(opts.Targets)
+		if err != nil {
+			return err
+		}
+
+		// get the manifests
+		res, err := LoadManifests(env, filter)
+		if err != nil {
+			return err
+		}
+
+		// create raw manifest version of env for templating
+		raw, err := json.Marshal(env)
+		if err != nil {
+			return err
+		}
+		var menv manifest.Manifest
+		if err := json.Unmarshal(raw, &menv); err != nil {
+			return err
+		}
+
+		// create template
+		manifestTemplate, err := createTemplate(*opts.Format, menv)
+		if err != nil {
+			return fmt.Errorf("Parsing format: %s", err)
+		}
+
+		// write each to a file
+		for _, m := range res {
+			// apply template
+			name, err := applyTemplate(manifestTemplate, m)
+			if err != nil {
+				return fmt.Errorf("executing name template: %w", err)
+			}
+
+			// Create all subfolders in path
+			path := filepath.Join(to, name+"."+*opts.Extension)
+
+			fileToEnv[path] = env.Metadata.Namespace
+
+			// Abort if already exists
+			if exists, err := fileExists(path); err != nil {
+				return err
+			} else if exists {
+				return fmt.Errorf("File '%s' already exists. Aborting", path)
+			}
+
+			// Write file
+			if err := os.MkdirAll(filepath.Dir(path), 0700); err != nil {
+				return fmt.Errorf("creating filepath '%s': %s", filepath.Dir(path), err)
+			}
+			data := m.String()
+			if err := ioutil.WriteFile(path, []byte(data), 0644); err != nil {
+				return fmt.Errorf("writing manifest: %s", err)
+			}
+		}
+	}
+
+	// create manifest file
+	if len(fileToEnv) != 0 {
+		data, err := json.MarshalIndent(fileToEnv, "", "    ")
+		if err != nil {
+			return err
+		}
+		path := filepath.Join(to, manifestFile)
+		if err := os.MkdirAll(filepath.Dir(path), 0700); err != nil {
+			return fmt.Errorf("creating filepath '%s': %s", filepath.Dir(path), err)
+		}
+		if err := ioutil.WriteFile(path, []byte(data), 0644); err != nil {
+			return err
+		}
+	}
```

Comment:
There is some duplicated bits that exist on the `cmd/tk/util.go:writeJSON` function, if we move that under `pkg/...` we could reuse it here. Something like: 

```go
                path := filepath.Join(to, manifestFile)
		if err := os.MkdirAll(filepath.Dir(path), 0700); err != nil {
			return fmt.Errorf("creating filepath '%s': %s", filepath.Dir(path), err)
		}
                 if err := <import>.WriteJSON(fileToEnv, path); err != nil {
			return err
		}
```

Reply:
I did a quick grep on `WriteFile` and found multiple very similar use cases. I propose consolidating that in a separate PR.

Reply:
I consolidated this bit in export.go, we can consolidate further in another PR.